### PR TITLE
auth, etcdserver: separate auth checking apply from core apply

### PR DESCRIPTION
**Pull request by heyitsanthony**
/cc @mitake @xiang90 

this is what I was suggesting in #5666 but it can wait until after the txn stuff is merged

### CODE REVIEW

**Comment by mitake** [comment]
@heyitsanthony thanks! I think adding permission checking of txn to this PR will be easier. I'll close https://github.com/coreos/etcd/pull/5666

File: etcdserver/apply_auth.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	}
+	aa.user = user
+	return aa.applierV3.Apply(r)
+}
+
+func (aa *authApplierV3) Put(txnID int64, r *pb.PutRequest) (*pb.PutResponse, error) {
+	if !aa.as.IsPutPermitted(aa.user, r.Key) {
+		return nil, auth.ErrPermissionDenied
+	}
+	return aa.applierV3.Put(txnID, r)
```
Comment (mitake) [inline etcdserver/apply_auth.go]: How about clearing `aa.user` after checking for ensuring the field is not used more than once?

**Comment by mitake** [comment]
@heyitsanthony lgtm (CI failure isn't related to the change)

File: etcdserver/apply_auth.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+func (aa *authApplierV3) DeleteRange(txnID int64, r *pb.DeleteRangeRequest) (*pb.DeleteRangeResponse, error) {
+	if !aa.as.IsDeleteRangePermitted(aa.user, r.Key, r.RangeEnd) {
+		return nil, auth.ErrPermissionDenied
+	}
+	return aa.applierV3.DeleteRange(txnID, r)
+}
+
+func (aa *authApplierV3) username(r *pb.InternalRaftRequest) (string, error) {
+	user := r.Header.Username
+	if needAdminPermission(r) && !aa.as.IsAdminPermitted(user) {
```
Comment (xiang90) [inline etcdserver/apply_auth.go]: i do not feel the admin checking should be instead the username func. move this to line32 right after calling username?

**Comment by xiang90** [comment]
LGTM

**Pull request merged**